File: python/dp/dp_b.py
#pypyなら通る
n, k = map(int, input().split())
h = [int(i) for i in input().split()]

dp = [0] * n
dp[1] = abs(h[1] - h[0])

for i in range(2,n):
    min_cost = 999999999
    for j in range(1,min(i,k)+1):
        tmp = dp[i-j] + abs(h[i] - h[i-j])
        min_cost = min(min_cost, tmp)
    dp[i] = min_cost

print(dp[n-1])

File: C++/dp_c++/dp_b.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0; i<(n); i++)
#define rep2(i,x,n) for(int i=x; i<(n); i++)
int main() {
    int N, K, tmp, min_cost;
    cin >> N >> K;
    vector<int> h(N), dp(N, 0);
    rep(i, N) cin >> h[i];
    dp[1] = abs(h[1] - h[0]);
    rep2(i, 2, N){
        min_cost = 999999999;
        rep2(j, 1, min(i, K)+1){
            tmp = dp[i-j] + abs(h[i] - h[i-j]);
            min_cost = min(min_cost, tmp);
        }
        dp[i] = min_cost;
    }
    cout << dp[N-1] << endl;
    return 0;
}